functions/attendance.js

```diff
@@ -0,0 +1,349 @@
+//
+// modified from: https://github.com/benclmnt/attendsys
+//
+
+const qrcode = require('qr.js');
+
+const { Client } = require('@notionhq/client');
+const notion = new Client({ auth: process.env.NOTION_API_KEY });
+
+// DEBUG is an environment variable
+
+const UNIQUE_CODE = '$CODE'
+const VALID_KEY = '$VALID_KEY'
+
+const NOTION_QR_CODE_PAGE_ID = process.env.NOTION_QR_CODE_PAGE_ID;
+
+async function updateAttendance(event) {
+  try {
+    const now = new Date()
+
+    if (!process.env.DEBUG) {
+      if (!isValidTimeRange(now) || !isValidDay(now)) {
+        throw new Error('Sorry, currently this service is unavailable')
+      }
+    }
+
+    const nusnet = escapeHtml(event.queryStringParameters.nusnet.toUpperCase())
+    const name = escapeHtml(event.queryStringParameters.name)
+    const code = event.path.substring(event.path.lastIndexOf("/") + 1)
+
+    if (code !== (await getUniqueCode())) {
+      throw new Error('Invalid attempt.')
+    }
+    
+    await markAttendance(nusnet, now)
+
+    // Redirect to /?name=${name}
+    url = event.path
+    url += '?name=' + name
+    return redirect(url)
+  } catch (err) {
+    return {
+      statusCode: 500,
+      body: "Error: \n" + err.message,
+    }
+  }
+}
+
+async function generateQRCode(event) {
+  const generatedUid = uid()
+  const urlWithCode =
+    'http:' + '//' + event.headers.host + event.path + generatedUid
+  console.log(urlWithCode)
+  const cells = qrcode(urlWithCode).modules
+
+  await updateUniqueCode(generatedUid)
+
+  return {
+      statusCode: 200,
+      body: qrTemplate(cells),
+      headers: { 'content-type': 'text/html' },
+  }
+}
+
+/**
+ * Respond with hello worker text
+ * @param {Request} request
+ */
+exports.handler = async function(event, context) {
+
+  // setting query parameter post=true as a workaround to avoid using POST
+  if (event.queryStringParameters.post != undefined) {
+    if (event.path.endsWith('/qr')) {
+      return generateQRCode(event)
+    }
+
+    return updateAttendance(event)
+  }
+
+  if (event.path.endsWith('/qr') || event.path === '/list') {
+    return {
+      statusCode: 200,
+      body: loginTemplate(),
+      headers: { 'content-type': 'text/html' },
+    }
+  }
+
+  const name = event.queryStringParameters.name
+  if (name) {
+    return {
+      statusCode: 200,
+      body: successTemplate(event.queryStringParameters.name),
+      headers: {
+          'content-type': 'text/html',
+      },
+    }
+  }
+
+  return {
+    statusCode: 200,
+    body: formTemplate(),
+    headers: {
+      'content-type': 'text/html'
+    },
+  }
+}
+
+/**
```

Comment:
I think the functions should not be returning the html. Instead, create a new page as part of the website, and let the function handles only the "update Notion" logic part. The page can then call this api / function.

Reply:
I'll move the html to its own proper page(s) 👍